rfcs/0014-vm-cycle-limits/0014-vm-cycle-limits.md

```diff
@@ -0,0 +1,111 @@
+
+This RFC describes cycle limits used to regulate VM scripts.
+
+CKB VM is a flexible VM that is free to implement many control flow constructs, such as loops or branches. As a result, we will need to enforce certain rules in CKB VM to prevent malicious scripts, such as a script with infinite loops.
+
+We introduce a concept called `cycles`, each VM instruction or syscall will consume some amount of cycles. At consensus level, a scalar `max_block_cycles` field is defined so that the sum of cycles consumed by all scripts in a block cannot exceed this value. Otherwise, the block will be rejected. This way we can guarantee all scripts running in CKB VM will halt, or result in error state.
+
+## Consensus Change
+
+As mentioned above, a new scalar `max_block_cycles` field is added to chain spec as a consensus rule, it puts a hard limit on how many cycles a block's scripts can consume. No block can consume cycles larger than `max_block_cycles`.
+
+Note there's no limit on the cycles for an individual transaction or a script. As long as the whole block consumes cycles less than `max_block_cycles`, a transaction or a script in that block are free to consume how many cycles they want.
+
+## Cycle Measures
+
+Here we will specify the cycles needed by each CKB VM instructions or syscalls. Note right now in the RFC, we define hard rules for each instruction or syscall here, in future this might be moved into consensus rules so we can change them more easily.
+
+### Instruction Cycles
+
+All CKB VM instructions consume 1 cycle except the following ones:
+
+| Instruction | Cycles |
+|-------------|--------|
+| JALR        | 3      |
+| JAL         | 3      |
+| J           | 3      |
+| JR          | 3      |
+| BEQ         | 3      |
+| BNE         | 3      |
+| BLT         | 3      |
+| BGE         | 3      |
+| BLTU        | 3      |
+| BGEU        | 3      |
+| BEQZ        | 3      |
+| BNEZ        | 3      |
+| LD          | 2      |
+| SD          | 2      |
+| LDSP        | 2      |
+| SDSP        | 2      |
+| LW          | 3      |
+| LH          | 3      |
+| LB          | 3      |
+| LWU         | 3      |
+| LHU         | 3      |
+| LBU         | 3      |
+| SW          | 3      |
+| SH          | 3      |
+| SB          | 3      |
+| LWSP        | 3      |
+| SWSP        | 3      |
+| MUL         | 5      |
+| MULW        | 5      |
+| MULH        | 5      |
+| MULHU       | 5      |
+| MULHSU      | 5      |
+| DIV         | 16     |
+| DIVW        | 16     |
+| DIVU        | 16     |
+| DIVUW       | 16     |
+| REM         | 16     |
+| REMW        | 16     |
+| REMU        | 16     |
+| REMUW       | 16     |
+| ECALL       | 0      |
+| EBREAK      | 0      |
+
+### Syscall Cycles
+
+Each syscall in CKB has different rules for consuming cycles:
+
+#### Load TX
+
+*Load TX* syscall first consumes 10 initial cycles, it then measures the size of the serialized transaction data: for every single byte in the data, it consumes 10 more cycles.
```

Comment:
Where do the numbers (10 initial and 10 for each byte) come from?

Reply:
This is just made-up number, the idea is syscalls should be relatively more expensive than normal instructions, but not too much (since syscalls are charged by data size).

The initial 10 comes from the fact that we want each syscall to at least consume some cycles since certain syscalls might return `ITEM_MISSING`, which means no data is available.